### ISSUE TITLE
Pinned items take too much UI space

### ISSUE THREAD

**Issue by borekb** [p0deje/Maccy#307]
I like to keep the number of items displayed at around 10 to keep the Maccy popup small and tidy. However, pinned items (if I'm not missing anything) cannot be limited or hidden from the main UI, which is a problem if I have many of them.

Also, while pinned items is a very useful feature of Maccy, I interact with it much less often than with the main history so it doesn't feel right that it should occupy the same space (or even more) in the UI.

Some ideas on how to solve this:

- Something like the mockup in https://github.com/p0deje/Maccy/issues/89#issuecomment-602265753 (a single "Pinned items" item in a list, opening a separate UI)
- Different shortcut and a dedicated UI to open pins, for example, like Clipy: <img width="641" alt="Screen Shot 2021-08-21 at 8 15 36" src="https://user-images.githubusercontent.com/101152/130312550-809855df-13fe-43f2-af58-aae0b1ec1e8b.png">

I'm sure there are other solutions, this is just to give an idea.

**Comment by p0deje** [comment]
Sorry for the slow response, I was away from Maccy for a while.

Since you don't interact with pinned items a lot, would it be useful for you to simply show them at the bottom so that the popup can be big but the pinned items won't take your focus?

**Comment by borekb** [comment]
I already have them at the bottom and I don't think it's enough.

Just to make sure, I'm cross-posting the mockup from [this comment](https://github.com/p0deje/Maccy/issues/89#issuecomment-602265753) which I think could be one of the solutions:

![image](https://user-images.githubusercontent.com/1752380/77259540-6d171900-6c82-11ea-980e-a1f76d7fcedc.png)

**Comment by greenwar** [comment]
tangent to this, i didn't find out to pin items from within the app itself. had to come to github via home site. it'd be great to add a blurb in the pins view in settings that tells the user how to pin. i could pull and update, but i wasn't sure how open this project is. cheers, love the app.

**Comment by WladyX** [comment]
I'm waiting for this feature to make the switch from clipy, i use lots of pinned items, so folders for pinned items would be perfect.
Thank you for this project.

**Comment by maliayas** [comment]
+1 for folders for pinned items (possibly hierarchical?). Only the top folder would be visible in the immediate popup UI. Going inside it would reveal the full tree of pins. The tree hierarchy editable via settings.

**Comment by douglasbeagley** [comment]
I would like to be able to create my own sub-menus with a set of saved pinned items.

For specific types of coding, html work, or when working in image editing with callouts, I like to keep a set of "frequently needed snippets" for that application or process. 

So, ideally I'd have both 1) the quick access to clipboard history (which Maccy does very well!), and 2) on the Maccy menu have a little submenu for "HTML Stuff" with the the fifteen specific pinned copies that I need, and one for "Callout Text", etc. Saved pins in categories.

**Comment by p0deje** [comment]
Are there any examples except for snippets when this could be useful? So far such a feature seems to be needed because people tend to use Maccy as a snippet manager and it's not a goal of the app. Adding abilities to nest and create hierarchical structures of pins looks a bit too much for a clipboard manager.

**Comment by tatablack** [comment]
I am also a Clipy user, and I use a few folders for pinned items there, for things that I'd like to be able to paste in different contexts. Some examples:
- my personal Zoom meeting link
- my bank details (account number and so on, nothing sensitive)
- symbols that I use when I'm writing, and that are not obvious to type (e.g. ™, ©, °, ⌘, ✓)
- sequences of characters that I'd never type from scratch, like `¯\_(ツ)_/¯` or `(╯°□°)╯︵ ┻━┻`

**Comment by olegcherr** [comment]
Many clipboard manages have pinned items. But what do you use pinned items for? I guess for frequently used items, like:
* emails
* phones
* addresses
* quick replies
* bank card numbers
* unusual symbols
etc.

You can see, than that's a quite big list of categories. Each category may contain about 3-5 items. So it's 20-30 items in total. That's too much for displaying on the main clipboard list.

1) So the first and easiest thing I would suggest - is the ability to place all the pinned items into a sub-menu. That would be much more convenient right away.

<img width="414" alt="image" src="https://github.com/p0deje/Maccy/assets/3730074/ae58d8d8-8683-4b67-a646-26ec12fce85a">

2) Second feature - is the ability to sort them in the order I need.

3) And the last thing - ability to group them:

<img width="447" alt="image" src="https://github.com/p0deje/Maccy/assets/3730074/c6f75c7d-b09d-472a-b22e-2f4b7a34a48b">

> Adding abilities to nest and create hierarchical structures of pins looks a bit too much for a clipboard manager.

I don't think so. Because in this case I need to find another app, that looks very very similar to Maccy, but allows to create snippets and paste them in any place. In other words, it should be Maccy, but only for snippets:))

And I don't want to find Maccy-for-snippets, because Maccy works better that most apps I've seen so far. So it would be awesome to see this functionality in Maccy.

Thanks!

**Comment by p0deje** [comment]
I understand the need for this feature, but I have yet to find a decent solution to all the extra complications from introducing folders:

1. How should the search work? Should it flatten the folders and show the matching results?
2. How would the keyboard navigation work?
3. What about Shortcuts actions that allow one to find items in history by their index?

**Comment by olegcherr** [comment]
> How should the search work? Should it flatten the folders and show the matching results?

I guess flattening folders would be a perfect solution. Found snippets should go last, after clipboard items, and maybe separated by divider.

> How would the keyboard navigation work?

Keyboard navigation should work like usual. Arrows up and down to go through items. Right and left arrows to enter and exit sub-menu. Cmd+<...> to paste a specific snippet.
![Untitled](https://github.com/p0deje/Maccy/assets/3730074/4fd2e2db-a37b-44b0-a136-12e916cd27e0)

> What about Shortcuts actions that allow one to find items in history by their index?

You mean these ones? I guess nothing should change here. They should work only for clipboard items, not for snippets.
<img width="415" alt="image" src="https://github.com/p0deje/Maccy/assets/3730074/4cc940d9-c8e8-4651-b694-6975967ab69d">

**Comment by p0deje** [comment]
> Keyboard navigation should work like usual. Arrows up and down to go through items. Right and left arrows to enter and exit sub-menu. Cmd+<...> to paste a specific snippet.

What about custom shortcuts that are assigned to pins? Should they still work even if the pinned item is not visible until folder is expanded?

> You mean these ones? I guess nothing should change here. They should work only for clipboard items, not for snippets.

No, I mean Shortcuts.app actions:
<img width="1312" alt="Screenshot 2024-05-27 at 15 18 27" src="https://github.com/p0deje/Maccy/assets/665846/dda15b5b-8a8b-4b26-a5fc-d011c45981ef">

**Comment by olegcherr** [comment]
> What about custom shortcuts that are assigned to pins? Should they still work even if the pinned item is not visible until folder is expanded?

Yes, sure. It would be the most convenient and intuitive solution.

> What about Shortcuts actions that allow one to find items in history by their index?

I don't know all the use cases of this Shortcut actions. But I think they should work with clipboard items only.

**Comment by kintaro1981** [comment]
I just started using Maccy and I was asking my self where "favorites" text were. 
I just added my 👍 to this feature request

**Comment by brettgil** [comment]
I am coming from Clipy as well. I appreciate your dedication and focus to keeping Maccy a clipboard manager focused.
The large number of PINs I want to keep will eat up a lot of real estate.  I understand the folders adds a lot of complexity. 1 simple option could be to just keep all PINs in a single fly out menu basically 1 folder with no hierarchy.  That would go a LONG way for me. Thanks. Keep up the great work.

**Comment by jantunes-te** [comment]
Just bumping this up (cc: @p0deje).
Maybe start with a simple implementation without folders or any complicated logic? It would "unblock" many of us waiting for this feature :D and you could always improve and iterate on it later.

**Comment by jantunes-te** [comment]
Just bumping this up (cc: @p0deje).
Maybe start with a simple implementation without folders or any complicated logic? It would "unblock" many of us waiting for this feature :D and you could always improve and iterate on it later.

**Comment by jasantunes** [comment]
Just bumping this up (cc: @p0deje).
Maybe start with a simple implementation without folders or any complicated logic? It would "unblock" many of us waiting for this feature :D and you could always improve and iterate on it later.
